File: Algorithm/leetcode/19.删除链表的倒数第-n-个结点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第 N 个结点
 */
// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	slow := fast = head
	for ; n--; {
		fast = fast.Next
	}
	if fast == nil {
		return head.Next
	}
	for ; fast.Next != nil && slow.Next != nil; {
		slow = slow.Next
		fast = fast.Next
	}
	slow.Next = slow.Next.Next
	return head
}
func getLength(head *ListNode) (length int) {
	for ; head != nil; head = head.Next {
			length++
	}
	return
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := getLength(head)
	dummy := &ListNode{0, head}
	cur := dummy
	for i := 0; i < length-n; i++ {
			cur = cur.Next
	}
	cur.Next = cur.Next.Next
	return dummy.Next
}
// @lc code=end

File: Algorithm/leetcode/links/19.删除链表的倒数第-n-个结点.js
/*
 * @lc app=leetcode.cn id=19 lang=javascript
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  // if(!head) return [];
  if (!head.next && n === 1) {
    return null;
  }
  let fast = slow = head;
  while (n--) {
    fast = fast.next;
  }
  if (!fast) {
    return head.next;
  }
  while (fast.next && slow.next) {
    fast = fast.next;
    slow = slow.next;
  }
  slow.next = slow.next.next;
  return head;
};
// @lc code=end

